src/otx/v2/adapters/torch/mmengine/mmseg/dataset.py

```diff
@@ -0,0 +1,252 @@
+            **kwargs: Additional arguments to pass to the DataLoader constructor.
+
+        Returns:
+            Optional[TorchDataLoader]: The DataLoader for the given dataset.
+        """
+        # TODO (Eugene): almost identical as mmengine.runner. Need to rethink about the design.
+        # CVS-124394
+
+        if dataset is None:
+            return None
+        rank, _ = get_dist_info()
+
+        # mmengine build dataloader in runner where we build in dataset.
+        # So, I'm basically copying things from runner.build_dataloader to here
+        # which does not make sense. why do we not use dataloader_cfg and build dataloader
+        # in runner.build_dataloader?
+        seed = kwargs.get("seed", 0)
+
+        # build sampler
+        if sampler is None:
+            sampler = DATA_SAMPLERS.build(
+                {"type": "DefaultSampler", "shuffle": shuffle},
+                default_args={"dataset": dataset, "seed": seed},
+            )
+
+        init_fn = partial(worker_init_fn, num_workers=num_workers, rank=rank, seed=seed) if seed is not None else None
+        if digit_version(torch.__version__) >= digit_version("1.8.0"):
+            kwargs["persistent_workers"] = persistent_workers
+
+        dataloader = TorchDataLoader(
+            dataset,
+            batch_size=batch_size,
+            sampler=sampler,
+            num_workers=num_workers,
+            collate_fn=partial(pseudo_collate),
+            pin_memory=pin_memory,
+            worker_init_fn=init_fn,
+            drop_last=drop_last,
+            **kwargs,
+        )
+        sampler_cfg = sampler if isinstance(sampler, dict) else {"type": f"{sampler.__class__.__qualname__}"}
+        dataset_cfg = dataset.configs if hasattr(dataset, "configs") else dataset
+        dataloader.configs = {
+            "batch_size": batch_size,
+            "sampler": sampler_cfg,
+            "num_workers": num_workers,
+            "collate_fn": {"type": "pseudo_collate"},
+            "pin_memory": pin_memory,
+            "shuffle": shuffle,
+            "dataset": dataset_cfg,
+        }
+        return dataloader
+
+    def _build_dataset(
+        self,
+        subset: str,
+        pipeline: list | None = None,
+        config: dict | None = None,
+    ) -> TorchDataset | None:
+        """Builds a TorchDataset object for the given subset using the specified pipeline and configuration.
+
+        Args:
+            subset (str): The subset to build the dataset for.
+            pipeline (list | None, optional): The pipeline to use for the dataset.
+                Defaults to None.
+            config (dict | None, optional): The configuration to use for the dataset.
+                Defaults to None.
+
+        Returns:
+            Optional[TorchDataset]: The built TorchDataset object, or None if the dataset is empty.
+        """
+        def check_and_convert_to_tuple(pipeline: list[dict] | None) -> list[dict] | None:
```

Comment:
don't we need this conversion for other frameworks when building dataset?

Reply:
If it is required only for mmseg, could you clarify what difference is between others?

Reply:
Each mmX has its own data pipeline transformation, so even mmdet might have a `RandomCrop` supports list, mmseg have a `RandomCrop` only allows `crop_size` being tuple.

Reply:
If so, how about doing this at src/otx/v2/adapters/torch/mmengine/mmseg/modules/datasets/otx_datasets.py? Some data preparation before augmentation may be performed at her. What do you think?

Reply:
no problem I will move this function there.

Reply:
thank you!